### ISSUE TITLE
Not able to render jupyter lab/notebooks in browser although the kernel is starting alright.

### ISSUE THREAD

**Issue by binarycache** [gitpod-io/openvscode-server#312]
- VS Code Version: 1.64.2 
- OS Version: 18.04.6 LTS

Steps to Reproduce:

1. Running the docker image from inside slurm accessed via ssh.
2. docker run -it --init -p 3000:3000 -v "$(pwd):/home/workspace:cached" gitpod/openvscode-server
3. Afte installing python kernel is getting selected.
4. Can't render ipynb files (Tried on different versions of python, even with conda environment)

Jupyter Log Output:
Info 15:40:06: Find preferred kernel for untitled:Untitled-1.ipynb?jupyter-notebook with metadata {"orig_nbformat":4,"language_info":{"name":"python"}} & preferred interpreter /usr/bin/python3
Info 15:40:06: findKernel found Python 3.6.9 64-bit
Info 15:40:06: PreferredConnection: .jvsc74a57bd031f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6./usr/bin/python3./usr/bin/python3.-m#ipykernel_launcher found for NotebookDocument: untitled:Untitled-1.ipynb?jupyter-notebook
Info 15:40:06: TargetController found ID: .jvsc74a57bd031f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6./usr/bin/python3./usr/bin/python3.-m#ipykernel_launcher for document untitled:Untitled-1.ipynb?jupyter-notebook
Info 15:40:06: Setting controller affinity for untitled:Untitled-1.ipynb?jupyter-notebook .jvsc74a57bd031f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6./usr/bin/python3./usr/bin/python3.-m#ipykernel_launcher
Info 15:40:06: Creating controller for jupyter-notebook with interpreter /usr/bin/python3

Regards,
Anmol

**Comment by nooobcoder** [comment]
That's surely a websocket and HTTPs issue, you need to setup a NGINX/Apache2 reverse proxy with SSL to load webviews in your browser. #96 This comment is very hot to solve your solution, although I could not find it helpful, as this does not explain from basics. And I wanted to setup a reverse proxy using docker.
If webviews are the problem, then this is the only solution for now. The jupyter extension in vscode uses webviews to render a notebook. @binarycache

**Comment by nooobcoder** [comment]
You can also hop into the Chrome dev tools and check for webviews not being loaded.
Thanks

**Comment by augustoicaro** [comment]
I had the same problem. The simple direct solution is to make a `ssh` with tunnel `-L` to slrum node machine to the port that openvscode-server was started (usually 3000), something like: `ssh -L 3000:slrum_node:3000 user@slrum_node`. And then, in your browser access as `localhost` (`localhost:3000`).

When running applications in localhost no HTTPS is required to activate the webworkers and the notebook will render as expected :)

**Comment by nooobcoder** [comment]
@augustoicaro I am using docker to host openvscode server, how do i achieve this?

**Comment by nooobcoder** [comment]
@augustoicaro Thanks it worked, i had no idea that SSH tunnels, fixes HTTPs issues.
This is not a permafix, but still would get my job easily done, thank you so much @augustoicaro

**Comment by augustoicaro** [comment]
@nooobcoder me too. Tell me about your scenario. Is it  a docker running in the same machine that you have the X and your browser is running?

**Comment by nooobcoder** [comment]
Yes @augustoicaro all works now, I have hosted it as a docker container and accessing remotely.
Thanks for your solution @augustoicaro, worked out like a charm!
Any mods reading this PIN his solution too or upvote as much possible.

**Comment by augustoicaro** [comment]
@nooobcoder I said the same thing that isn't a permafix, but was never prioritized since that tunnel solution is working well...

**Comment by binarycache** [comment]
@augustoicaro Worked liked a charm! @nooobcoder 
Note: You might need to do ssh -L 3000:ip_address_ssh_server:3000 user@ssh_server instead of using the name of the server for port tunneling. 
P.S. I still got an error of webview permissions which was basically a problem with "brave's" shields, try turning off the shields, and allowing cookies and javacript. I switched to chrome though for quick testing.
P.P.S I am still getting an error saying couldn't start python client tools which might be an issue with language server but it's working fine.

Thanks again!

**Comment by nooobcoder** [comment]
@binarycache that was very predictable. You can either use the FQDN or the IP address of the server whose port you want to forwarded to your local host.
You can read more about SSH port forwarding on: [linuxize](https://linuxize.com/post/how-to-setup-ssh-tunneling/)
Good day.

**Comment by nooobcoder** [comment]
@binarycache 

`P.P.S I am still getting an error saying couldn't start python client tools which might be an issue with language server but it's working fine.`

Please provide further logs to track this down

**Comment by binarycache** [comment]
@nooobcoder I am not sure how to get the logs from it but the issue is and its solution is mentioned here https://github.com/Microsoft/vscode-python/issues/2877

**Comment by nooobcoder** [comment]
@jeanp413 can i have a easy guide with Docker compose to have a reverse proxy with websockets for openvscode server. I want a least brainf**k method of rev proxy using docker.

**Comment by augustoicaro** [comment]
> @augustoicaro Worked liked a charm! @nooobcoder Note: You might need to do ssh -L 3000:ip_address_ssh_server:3000 user@ssh_server instead of using the name of the server for port tunneling. P.S. I still got an error of webview permissions which was basically a problem with "brave's" shields, try turning off the shields, and allowing cookies and javacript. I switched to chrome though for quick testing. P.P.S I am still getting an error saying couldn't start python client tools which might be an issue with language server but it's working fine.
> 
> Thanks again!

About python language service, it will not work because it is not installed with Microsoft python extension, due to proprietary restrictions as remote extensions. So I use `Jedi` as language service in my projects.
